src/Stiefel.jl

```diff
@@ -0,0 +1,199 @@
+using LinearAlgebra: diag, qr, tr, svd, mul!, lyap
+
+@doc doc"""
+    Stiefel{m,n,T} <: Manifold
+
+The Stiefel manifold consists of all $m\times n$, $m\geq n$ orthonormal matrices, i.e.
+
+````math
+\mathcal M = \{ x \in \mathbb K^{m\times n} : \overline{x}^{\mathrm{T}}x = I_n \}
+````
+where $\mathbb K \in \{\mathbb R, \mathbb C\}$ and
+$I_n \in \mathbb R^{n\times n}$ denotes the unit matrix.
+
+The tangent space at a point $x\in\mathcal M$ is given by
+````math
+T_x\mathcal M = \{ v \in \mathbb{K}^{m\times n} : x^{\mathrm{T}}v + v^{\mathrm{T}}x=0_n\},
+````
+where $0_n$ is the $n\times n$ zero matrix.
+
+# Constructor
+    Stiefel(m,n,T=Real)
+
+generate the (real-valued) Stiefel manifold of $m\times n$ dimensional orthonormal matrices.
+"""
+struct Stiefel{M,N,T} <: Manifold end
+Stiefel(m::Int, n::Int,T::Type = Type{Real} ) = Stiefel{m,n,T}()
+
+function check_manifold_point(S::Stiefel{M,N,T},x; kwargs...) where {M,N,T}
+    if any( size(x) != representation_size(S) )
+        return DomainError(size(x),
+            "The matrix $(x) is does not lie on the Stiefel manifold of dimension (m,n), since its dimensions are wrong.")
+    end
+    c = x'*x
+    if !isapprox(c, one(c); kwargs...)
+        return DomainError(norm(c-one(c)),
+            "The point $(x) does not lie on the Stiefel manifold of dimension (m,n), because x'x is not the unit matrix.")
+    end
+end
+
+function check_tangent_vector(S::Stiefel{M,N,T},x,v; kwargs...) where {M,N,T}
+    t = check_manifold_point(S,x)
+    if (t != nothing)
+        return t
+    end
+    if any( size(v) != representation_size(S) )
+        return DomainError(size(v),
+            "The matrix $(v) is does not lie in the tangent space of $(x) on the Stiefel manifold of dimension (m,n), since its dimensions are wrong.")
+    end
+    if !isapprox(x'*v + v'*x, zeros(N,N); kwargs...)
+        return DomainError(norm(x'*v + v'*x),
+            "The matrix $(v) is does not lie in the tangent space of $(x) on the Stiefel manifold of dimension (m,n), since x'v + v'x is not the zero matrix.")
+    end
+end
+@doc doc"""
+    exp!(M,y, x, v)
+
+compute the exponential map on the [`Stiefel`](@ref)`{m,n,T}`() manifold `M`
+eminating from `x` into tangent direction `v` and store the result in `y`.
+The formula reads
+
+$\operatorname{exp}_{x} v = \begin{pmatrix}
+   x\\v
+ \end{pmatrix}
+ \exp
+ \left(
+ \begin{pmatrix} {\bar x}^{\mathrm{T}}v & -{\bar \xi}^{\mathrm{T}}v\\
+ I_n & {\bar x}^{\mathrm{T}}v\end{pmatrix}
+ \right)
+\begin{pmatrix}  \exp( -{\bar x}^{\mathrm{T}}v) \\ 0_n\end{pmatrix}$
+
+where $\exp$ denotes matrix exponential, and $I_n$ and
+$0_n$ are the identity matrix and the zero matrix of dimension $n \times n$,
+respectively.
+"""
+function exp!(M::Stiefel{m,n,T}, y, x, v) where {m,n,T}
+    y .= [x v] * exp([x'v-v'*v; one(zeros(T,n,n)) x'*v]) * [exp(-x'v); zeros(T,n,n)]
```

Comment:
I really don't like this line but it's a tough one to fix. @dkarrasch , could you take a look at this?